Add DPT Flax

younesbelkada opened the pull request:
# What does this PR do?

I tried to implement DPT (Dense Prediction with Transformers) in Flax during my free time! 🚀 
By the way it is the first Segmentation and Depth Estimation model implemented in Flax on the library!

Nits/TODOs:
- [x] Figure out how to properly call `BatchNorm` and `Dropout` inside a `Sequential` 
- [x] Deal correctly with `Sequential` layers 
- [x] Test equivalency tests 
- [ ] Write documentation - For now they're just copy/pasted
Quetions:
- Why the loss is not implemented in `modeling_dpt.py` ? I can probably help on that since I have already implemented the loss for a university project: https://github.com/antocad/FocusOnDepth/blob/master/FOD/Loss.py 
 
cc @NielsRogge @sanchit-gandhi @patil-suraj

younesbelkada committed:
first commit:

- added DPT in Flax
- all non slow tests passes in local
- still some nits have to be investigated

younesbelkada committed:
make quality

ArthurZucker committed:
implement pt_flax equivlence test to bypass  `AttributeError: 'NoneType' object has no attribute 'tolist'`

younesbelkada committed:
add few changes:

- BN seems to work now
- Equivalency test pass with tol=1e-4 but only with a hack

younesbelkada committed:
add modification sequential

ArthurZucker commented on src/transformers/models/dpt/modeling_flax_dpt.py:
```diff
@@ -0,0 +1,1231 @@
+    - [Vectorization](https://jax.readthedocs.io/en/latest/jax.html#vectorization-vmap)
+    - [Parallelization](https://jax.readthedocs.io/en/latest/jax.html#parallelization-pmap)
+    Parameters:
+        config ([`ViTConfig`]): Model configuration class with all the parameters of the model.
```
`DPTConfig` instead of ViT

younesbelkada replied:
Fixed in 6b2c28c81b8424ca8435425f6016670c3eafb884

HuggingFaceDocBuilderDev commented:
The docs for this PR live [here](https://moon-ci-docs.huggingface.co/docs/transformers/pr_17779). All of your documentation changes will be reflected on that endpoint.

younesbelkada committed:
fix copies

ArthurZucker commented on src/transformers/models/dpt/modeling_flax_dpt.py:
```diff
@@ -0,0 +1,1237 @@
+            self.config.fusion_hidden_size // 2, kernel_size=(3, 3), strides=(1, 1), padding=1, name="0"
+        )
+
+        self.upsample = FlaxDPTUpsample(scale=2, method="bilinear")
```
Align corner is the source of the issue

ArthurZucker commented on src/transformers/models/dpt/modeling_flax_dpt.py:
```diff
@@ -0,0 +1,1237 @@
+        return output
+
+
+class FlaxDPTUpsample(nn.Module):
+    scale: int = 2
+    method: str = "bilinear"
+
+    def setup(self):
+        pass
+
+    def __call__(self, x, output_size=None):
+        if output_size is None:
+            output_size = x.shape
+            output_size = (output_size[0], output_size[1] * self.scale, output_size[2] * self.scale, output_size[3])
+        return jax.image.resize(x, output_size, method="bilinear")
```
Should support `align_corners = True`

younesbelkada committed:
few fixes

- more documentation
- fix nit

NielsRogge commented:
Would be great to also incorporate the updates of #17731

younesbelkada commented on src/transformers/models/dpt/modeling_flax_dpt.py:
```diff
@@ -0,0 +1,1237 @@
+
+    def setup(self):
+        self.cls_token = self.param("cls_token", nn.initializers.zeros, (1, 1, self.config.hidden_size))
+        self.patch_embeddings = FlaxPatchEmbeddings(self.config, dtype=self.dtype)
```
@NielsRogge I think that in this implementation we directly initialize the modules with the config (contrary than in `DPTViTEmbeddings` if this is what you meant?

younesbelkada replied:
I copied these modules from `FlaxViTModel` which seem to have the right structure as suggested in #17731

NielsRogge replied:
Ok, then it's alright :)

younesbelkada replied:
Oh but the sanity check of the channel size is missing indeed, will add that!

NielsRogge replied:
Yeah we don't have it for other Flax models as well right now. Ideally (and also for consistency), we should have it for all vision models

younesbelkada committed:
changes

- add custom conv transpose2d function
- modify test

younesbelkada commented:
The Flax model finally predicts the correct depths for the cats (left is Flax and right is Pytorch)! 
![Screenshot 2022-06-25 at 20 12 08](https://user-images.githubusercontent.com/49240599/175785580-bd9b3b25-cc90-4ad0-a262-fbf335e1351f.png)
For that it appears that the transpose conv does not give the same result as Pytorch's implementation that uses a gradient based operation. I fixed it by creating a custom function based on this PR: https://github.com/google/jax/pull/5772 the PR does not seem to be merged soon. We can probably go for this hack for now until the PR gets merged

ArthurZucker committed:
Merge branch 'main' of https://github.com/huggingface/transformers into dpt-flax-younes

ArthurZucker committed:
style

ArthurZucker committed:
add FlaxViTPatchEmbeddings for consistency

ArthurZucker committed:
update tests

ArthurZucker committed:
consistency

ArthurZucker committed:
style

ArthurZucker commented:
As wee discussed, it seems that `align_corners` set to  `False` for both model would not require lowering the tolerance in one of the cases right?

younesbelkada commented:
@ArthurZucker exact. I have put a new attribute in the `DPTConfig` and modified a bit the original modeling code but should not break backward compatibility. Now all tests pass with a tolerance of `1e-5`

younesbelkada committed:
all tests should pas

- added new attribute to config without breaking backward compatibility
- modified a bit the tests

sanchit-gandhi commented on src/transformers/models/dpt/gradient_convolution.py:
```diff
@@ -0,0 +1,451 @@
+def _flip_axes(x, axes):
+    """Flip ndarray 'x' along each axis specified in axes tuple."""
+    for axis in axes:
+        x = np.flip(x, axis)
```
```suggestion
        x = jnp.flip(x, axis)
```

younesbelkada replied:
Should be fixed but I forgot to commit your suggestion 😭 Sorry for that!

younesbelkada replied:
See 7c26f69f264305338e1fc285ecd8b4c8e3aeaf38

sanchit-gandhi commented on src/transformers/models/dpt/gradient_convolution.py:
```diff
@@ -0,0 +1,451 @@
+            raise ValueError("No 4+ dimensional dimension_number defaults.")
+    dn = conv_dimension_numbers_(lhs.shape, rhs.shape, dimension_numbers)
+    # dn = dimension_numbers
+    k_shape = np.take(rhs.shape, dn.rhs_spec)
```
I would advocate for using `jnp` calls instead of `np`. This function is going to be jit’d, so we want to keep everything on the accelerator device where possible. `np` calls that pull variables back to the host.

younesbelkada replied:
I found this a bit tricky, moving it to `jnp` calls will implicate going through [this statement](https://github.com/younesbelkada/transformers/blob/a40f21dfe73079f26c5afccccad1e1cf14d15ac8/src/transformers/models/dpt/gradient_convolution.py#L208): 
```
if expected_input_size != input_size,:
            raise ValueError(
                "The expected input size with the current set of input "
                f"parameters is {expected_input_size} which doesn't "
                f"match the actual input size {input_size}."
            )
``` 
Which I found a bit hard to refactor to make it run under `jit`: https://jax.readthedocs.io/en/latest/errors.html#jax.errors.ConcretizationTypeError 
Do you have any idea how we can go around this statement 🙏 ?

sanchit-gandhi replied:
Could try with a `if not jnp.equal(expected_input_size, input_size):`

sanchit-gandhi commented on src/transformers/models/dpt/gradient_convolution.py:
```diff
@@ -0,0 +1,451 @@
+
+
+class ConvTransposeGradient(nn.Module):
+    """Convolution Module wrapping lax.conv_transpose.
```
Could also say a word about how this is borrowed largely from [`flax.linen.ConvTranspose`](https://flax.readthedocs.io/en/latest/_modules/flax/linen/linear.html#ConvTranspose) and how it differs.
```suggestion
    """Convolution Module wrapping lax.conv_transpose. Calculates transposed convolutions via gradient (transpose) of a forward convolution.
```

sanchit-gandhi commented on src/transformers/models/dpt/gradient_convolution.py:
```diff
@@ -0,0 +1,451 @@
+    padding: PaddingLike = "SAME"
+    kernel_dilation: Optional[Sequence[int]] = None
+    use_bias: bool = True
+    dtype: Dtype = jnp.float32
```
For `PRNGKey`, `Dtype` and `Array`, IMO it's more readable to include the actual type hint in each class definition rather than a variable that we define here
```suggestion
    dtype: jnp.dtype = jnp.float32
```

sanchit-gandhi commented on src/transformers/models/dpt/gradient_convolution.py:
```diff
@@ -0,0 +1,451 @@
+    kernel_dilation: Optional[Sequence[int]] = None
+    use_bias: bool = True
+    dtype: Dtype = jnp.float32
+    param_dtype: Dtype = jnp.float32
```
We usually fix the `param_dtype` to `jnp.float32` and omit the flag from our scripts: it tends to induce clutter including it in every module definition.

sanchit-gandhi commented on src/transformers/models/dpt/gradient_convolution.py:
```diff
@@ -0,0 +1,451 @@
+    kernel_init: Callable[[PRNGKey, Shape, Dtype], Array] = default_kernel_init
+    bias_init: Callable[[PRNGKey, Shape, Dtype], Array] = zeros
+
+    @compact
```
We tend to avoid defining ‘compact’ `nn.Module` classes: we believe including the `setup` makes them more readable. But if we make clear that this is largely lifted from [`flax.linen.ConvTranspose`](https://flax.readthedocs.io/en/latest/_modules/flax/linen/linear.html#ConvTranspose) we can avoid confusion

younesbelkada replied:
Yeah I think that adding a comment explaining that the module is lifted from https://flax.readthedocs.io/en/latest/_modules/flax/linen/linear.html#ConvTranspose
Should be fixed in 24aeb4d13a735ad11241736233b26a80ba203d35

younesbelkada replied:
Added more details in f945eef63b733bd247b0d5d7c83a9fa28f06467c

sanchit-gandhi commented on src/transformers/models/dpt/gradient_convolution.py:
```diff
@@ -0,0 +1,451 @@
+    def __call__(self, inputs: Array) -> Array:
+        """Applies a transposed convolution to the inputs.
+
+        Behaviour mirrors of `jax.lax.conv_transpose`.
```
```suggestion
        Behaviour mirrors of `jax.lax.conv_transpose`, computing transposed convolutions via the gradient (transpose) of a forward convolutions.
```

sanchit-gandhi commented on src/transformers/models/dpt/modeling_flax_dpt.py:
```diff
@@ -0,0 +1,1236 @@
+        outputs = (attn_output, attn_weights) if output_attentions else (attn_output,)
+        return outputs
+
+
+class FlaxDPTViTOutput(nn.Module):
```
```suggestion
`# Copied from transformers.models.vit.modeling_flax_vit.FlaxViTOutput with ViT->DPT`
class FlaxDPTViTOutput(nn.Module):
```

sanchit-gandhi commented on src/transformers/models/dpt/modeling_flax_dpt.py:
```diff
@@ -0,0 +1,1236 @@
+    dtype: jnp.dtype = jnp.float32
+
+    def setup(self):
+        self.dense = nn.Dense(self.config.hidden_size, name="0")
```
Do we need to define name? Will be overwritten by `name=str(i)` in L337

younesbelkada replied:
I think that I had to put it because we need the dense layer to have the key value `0` to match the state dict of Pytorch.
Since in pytorch the `FlaxDPTReadoutProjectSequentialCollectionLayer` are `nn.Sequential` layers the first layer will have the key `0` -> see [this](https://github.com/younesbelkada/transformers/blob/010a915ce62bd1a3adb61d0eab2f0acca436b1c0/src/transformers/models/dpt/modeling_dpt.py#L438) line where it happens in Pytorch. That is also why the class is called `FlaxDPTReadoutProjectSequentialCollectionLayer`. 
Will add a comment to explain why we need this name

younesbelkada replied:
Added more comments at: a40f21dfe73079f26c5afccccad1e1cf14d15ac8

sanchit-gandhi commented on src/transformers/models/dpt/modeling_flax_dpt.py:
```diff
@@ -0,0 +1,1236 @@
+        if self.config.readout_type == "project":
+            self.readout_projects = FlaxDPTReadoutProjectCollectionLayer(self.config, self.dtype)
+
+    def __call__(self, hidden_states):
```
This `__call__` method is nice!

younesbelkada replied:
Thanks!  :)

sanchit-gandhi commented on src/transformers/models/dpt/modeling_flax_dpt.py:
```diff
@@ -0,0 +1,1236 @@
+class FlaxDPTFeatureFusionLayer(nn.Module):
+    config: DPTConfig
+    dtype: jnp.dtype = jnp.float32
+    align_corners: bool = True
```
`align_corners` defined but not used. Should this attribute override the kwarg `size` in the `__call__` method for the `nn.Module` `FlaxDPTUpsample`?

younesbelkada replied:
I think that I messed up something there, we should not need `align_corners` as you said. I have refactored it here: 791ea245ed3470463d3a9e6fee2c49d384a4a71a

sanchit-gandhi commented on src/transformers/models/dpt/modeling_flax_dpt.py:
```diff
@@ -0,0 +1,1236 @@
+    align_corners: bool = True
+
+    def setup(self):
+        self.projection = nn.Conv(self.config.fusion_hidden_size, kernel_size=(1, 1))  # , bias=True)
```
`use_bias` defaults to `True`, but can pass`use_bias=True` as a kwarg for clarity (as done in PT)
```suggestion
        self.projection = nn.Conv(self.config.fusion_hidden_size, kernel_size=(1, 1), use_bias=True)
```

sanchit-gandhi commented on src/transformers/models/dpt/modeling_flax_dpt.py:
```diff
@@ -0,0 +1,1236 @@
+        if residual is not None:
+            if hidden_state.shape != residual.shape:
+                size = hidden_state.shape
+                residual = self.upsample(residual, size)
```
Set  `align_corners=True`
```suggestion
                residual = self.upsample(residual, align_corners=True)
```

younesbelkada replied:
Fixed in 791ea245ed3470463d3a9e6fee2c49d384a4a71a

sanchit-gandhi commented on src/transformers/models/dpt/modeling_flax_dpt.py:
```diff
@@ -0,0 +1,1236 @@
+            hidden_state = hidden_state + self.residual_layer1(residual)
+
+        hidden_state = self.residual_layer2(hidden_state)
+        hidden_state = self.upsample(hidden_state)
```
Set `align_corners=self.align_corners` or `align_corners=self.config.align_corners`
```suggestion
        hidden_state = self.upsample(hidden_state, align_corners=self.align_corners)
```

younesbelkada replied:
Fixed in 791ea245ed3470463d3a9e6fee2c49d384a4a71a

sanchit-gandhi commented on src/transformers/models/dpt/modeling_flax_dpt.py:
```diff
@@ -0,0 +1,1236 @@
+
+        return hidden_state
+
+
+class FlaxDPTViTLayer(nn.Module):
```
```suggestion
# Copied from transformers.models.vit.modeling_flax_vit.FlaxViTLayer with ViTConfig->DPTConfig, ViTAttention->DPTViTAttention, ViTIntermediate->DPTViTIntermediate, ViTOutput->DPTViTOutput
class FlaxDPTViTLayer(nn.Module):
```

sanchit-gandhi commented on src/transformers/models/dpt/modeling_flax_dpt.py:
```diff
@@ -0,0 +1,1236 @@
+            outputs += (attention_outputs[1],)
+        return outputs
+
+
+class FlaxDPTViTLayerCollection(nn.Module):
```
```suggestion
# Copied from transformers.models.vit.modeling_flax_vit.FlaxViTLayerCollection with ViTConfig->DPTConfig, ViTLayer->DPTViTLayer
class FlaxDPTViTLayerCollection(nn.Module):
```

younesbelkada replied:
This does seem to not work in : 791ea245ed3470463d3a9e6fee2c49d384a4a71a will investigate that!

younesbelkada replied:
Actually I think that we cannot copy it from `FlaxViTLayerCollection` since to do properly the forward pass you would need all the hidden states for each stage. On `FlaxViTLayerCollection` the tuple `all_hidden_states` is not returned which is what we use [here](https://github.com/younesbelkada/transformers/blob/f945eef63b733bd247b0d5d7c83a9fa28f06467c/src/transformers/models/dpt/modeling_flax_dpt.py#L1010) at the index 1

sanchit-gandhi commented on src/transformers/models/dpt/modeling_flax_dpt.py:
```diff
@@ -0,0 +1,1236 @@
+            last_hidden_state=hidden_states, hidden_states=all_hidden_states, attentions=all_attentions
+        )
+
+
+class FlaxDPTViTPooler(nn.Module):
```
```suggestion
# Copied from transformers.models.vit.modeling_vit.FlaxViTPooler with ViT->DPT
class FlaxDPTViTPooler(nn.Module):
```

sanchit-gandhi commented on src/transformers/models/dpt/modeling_flax_dpt.py:
```diff
@@ -0,0 +1,1236 @@
+        cls_hidden_state = self.dense(cls_hidden_state)
+        return nn.tanh(cls_hidden_state)
+
+
+class FlaxDPTViTEncoder(nn.Module):
```
```suggestion
# Copied from transformers.models.vit.modeling_flax_vit.FlaxViTEncoder with ViTConfig -> DPTConfig, ViTLayer->DPTViTLayer
class FlaxDPTViTEncoder(nn.Module):
```

sanchit-gandhi commented on src/transformers/models/dpt/modeling_flax_dpt.py:
```diff
@@ -0,0 +1,1236 @@
+        if input_shape is None:
+            input_shape = (1, config.image_size, config.image_size, 3)
+        super().__init__(config, module, input_shape=input_shape, seed=seed, dtype=dtype, _do_init=_do_init)
+
+    def init_weights(self, rng: jax.random.PRNGKey, input_shape: Tuple, params: FrozenDict = None) -> FrozenDict:
```
```suggestion
    # Copied from transformers.models.FlaxViTPreTrainedModel.init_weights
    def init_weights(self, rng: jax.random.PRNGKey, input_shape: Tuple, params: FrozenDict = None) -> FrozenDict:
```

sanchit-gandhi commented on src/transformers/models/dpt/modeling_flax_dpt.py:
```diff
@@ -0,0 +1,1236 @@
+            return random_params
+
+    @add_start_docstrings_to_model_forward(DPT_INPUTS_DOCSTRING.format("batch_size, sequence_length"))
+    def __call__(
```
```suggestion
    # Copied from transformers.models.FlaxViTPreTrainedModel.__call__
    def __call__(
```

sanchit-gandhi commented on src/transformers/models/dpt/modeling_flax_dpt.py:
```diff
@@ -0,0 +1,1236 @@
+    def setup(self):
+        pass
+
+    def __call__(self, x, output_size=None):
+        if output_size is None:
```
Wonder if we can improve clarity here by aligning with PT?
```suggestion
    def __call__(self, x, align_corners=True):
        if not align_corners:
```

sanchit-gandhi reviewed (changes requested):
Hey @younesbelkada! Looks pretty good from the Flax side of things! Left a few requests, the overarching one being the use of `# Copied from...` statements, both for internal Transformers code and that copied externally (e.g. from Haiku). Really helps in knowing what are the salient portions of the modelling code to review! But otherwise a very strong effort on the Flax front 💪

younesbelkada committed:
fixing few comments

younesbelkada commented on src/transformers/models/dpt/gradient_convolution.py:
```diff
@@ -0,0 +1,451 @@
+    dilation: int = 1,
+) -> Tuple[int, int]:
+    """Computes adjusted padding for desired ConvTranspose `output_size`.
+    Ported from DeepMind Haiku.
```
I just added the correct reference to the original file ;) 7c26f69f264305338e1fc285ecd8b4c8e3aeaf38 Thanks for the pointer!

younesbelkada committed:
Apply suggestions from code review

Co-authored-by: Sanchit Gandhi <93869735+sanchit-gandhi@users.noreply.github.com>

younesbelkada commented on src/transformers/models/dpt/gradient_convolution.py:
```diff
@@ -0,0 +1,451 @@
+
+PRNGKey = Any
+Shape = Tuple[int, ...]
+Dtype = Any  # this could be a real type?
```
Thanks for the catch! Should be fixed in 7c26f69f264305338e1fc285ecd8b4c8e3aeaf38

younesbelkada commented on src/transformers/models/dpt/gradient_convolution.py:
```diff
@@ -0,0 +1,451 @@
+Array = Any
+PrecisionLike = Union[None, str, lax.Precision, Tuple[str, str], Tuple[lax.Precision, lax.Precision]]
+
+PrecisionLike = Union[None, str, lax.Precision, Tuple[str, str], Tuple[lax.Precision, lax.Precision]]
```
Totally! Should be fixed in 7c26f69f264305338e1fc285ecd8b4c8e3aeaf38

younesbelkada committed:
Update src/transformers/models/dpt/gradient_convolution.py

Co-authored-by: Sanchit Gandhi <93869735+sanchit-gandhi@users.noreply.github.com>

younesbelkada commented on src/transformers/models/dpt/configuration_dpt.py:
```diff
@@ -137,6 +137,7 @@ def __init__(
         auxiliary_loss_weight=0.4,
         semantic_loss_ignore_index=255,
         semantic_classifier_dropout=0.1,
+        align_corners=True,
```
Great !! Thanks ;)

younesbelkada committed:
add few comments

younesbelkada committed:
Apply suggestions from code review

Co-authored-by: Sanchit Gandhi <93869735+sanchit-gandhi@users.noreply.github.com>

younesbelkada committed:
refactor a bit:

- removed `copied_from` on non module objects
- check why `FlaxDPTViTLayerCollection` is not copied from `FlaxViTLayerCollection`

younesbelkada committed:
Merge remote-tracking branch 'upstream/main' into dpt-flax-younes

younesbelkada committed:
add comments on key naming strategy

younesbelkada committed:
few modifications

- added correct link for `CopiedFrom`
- Added explicit argument for transposed conv on model def

younesbelkada commented:
Thank you very much @sanchit-gandhi for the very detailed review! I had a second round of refactoring while catching up on Flax projects and would love to have a second round of review 💪 Thanks again 🙏
